cmd/server.go

```diff
@@ -20,44 +20,134 @@ import (
 	"context"
 	"crypto/tls"
 	"fmt"
+	dockhand "github.com/boxboat/dockhand-secrets-operator/pkg/apis/dockhand.boxboat.io/v1alpha1"
 	"github.com/boxboat/dockhand-secrets-operator/pkg/common"
 	"github.com/boxboat/dockhand-secrets-operator/pkg/k8s"
 	"github.com/boxboat/dockhand-secrets-operator/pkg/webhook"
 	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/leaderelection"
 	"net/http"
 	"os"
 	"os/signal"
 	"syscall"
+	"time"
 )
 
 type ServerArgs struct {
-	serverPort        int
-	serverCert        string
-	serverKey         string
-	serviceName       string
-	serviceNamespace  string
-	selfSignCerts     bool
+	serverPort       int
+	serverCert       string
+	serverKey        string
+	serviceName      string
+	serviceId        string
+	serviceNamespace string
+	selfSignCerts    bool
 }
 
 var (
 	serverArgs ServerArgs
 )
 
-func runServer(ctx context.Context) {
+func certManager() {
+	lock, err := k8s.GetLeaseLock(serverArgs.serviceId, serverArgs.serviceName, serverArgs.serviceNamespace)
+	common.ExitIfError(err)
+
+	leaderCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	leaderelection.RunOrDie(leaderCtx, leaderelection.LeaderElectionConfig{
+		Lock:          lock,
+		LeaseDuration: 15 * time.Second,
+		RenewDeadline: 10 * time.Second,
+		RetryPeriod:   2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				var cert *tls.Certificate
+				checkTime := time.Now()
+				for {
+					// only retrieve the certificate at startup and then once a day after that
+					if cert == nil || time.Now().Add(time.Hour*-24).After(checkTime) {
```

Comment:
Might not hurt to check for the cert more frequently in the event the secret is somehow deleted. Maybe at least make the duration configurable?

Reply:
This is pretty much the only issue I see also.  Secret disappears from under the leader, then another replica tries to startup, but crash loops until 24h later when the secret is re-generated.  Definitely an edge case.  Might be able to watch the secret for updates / deletion, not sure how hard that is

Reply:
Changed to 1m